### PR TITLE
Build wheel using cibuildwheel

### LINKED ISSUE #93: Could you upload python3.7 python3.8 wheel to pypi.org?
RT

Thank you.

### PR DESCRIPTION
**Pull request by Czaki**
Closes #93 

This PR build wheel for Linux and macOS. 
Its build wheel for CPython 2.7, 3.5-9. PyPy 2.7, 3.6, 3.7 for Linux 32 bits/64 bits and macOS. 

It builds CPython 3.5-9 pypy3.6, pypy3.7 for windows. 

When `cibuilwheel` 2.0 was released then also macOS `universal2` could be a simple build.  

This PR move source code to `src` to avoid name collision when executing tests.

### CODE REVIEW

**Comment by martindurant** [comment]
I can't personally help with the Windows thing. Perhaps someone else can.

**Comment by Czaki** [comment]
@martindurant I read a little and it is now building windows wheel above 3.5.

**Comment by Czaki** [comment]
@martindurant I'm an active contributor to `cibuildwheel` so if You have any questions I could try to explain some nonclear parts.

**Comment by martindurant** [comment]
What would trigger an actual upload to pypi, a new git tag? 

I notice that a lot of files were moved around such as snappy/snappy.py → src/snappy/snappy.py; this, for me, beraks the typical pydata package layout, so I wonder whether it is necessary.

File: .github/workflows/wheel.yml [hunk 1]
```diff
@@ -1,79 +1,127 @@
...
+      - uses: actions/upload-artifact@v2
+        with:
+          path: ./wheelhouse/*.whl
+          name: wheels
+
+  upload:
+    runs-on: ubuntu-latest
+    name: upload wheels
+    needs: ['sdist', 'build']
+    if: startsWith(github.ref, 'refs/tags/0.')
```
Comment (Czaki) [inline .github/workflows/wheel.yml:116]: here is upload trigger

**Comment by Czaki** [comment]
> I notice that a lot of files were moved around such as snappy/snappy.py → src/snappy/snappy.py; this, for me, beraks the typical pydata package layout, so I wonder whether it is necessary.

this is because the test grabs a snappy folder instead of the content of the snappy package. The previous workflow delete files to prevent this.  

I strongly prefer `src` directory, to prevent PYTHONPATH pollution. 

Maybe the other solution could be a move test (which also should not be in a top lever directory).

File: .github/workflows/wheel.yml [hunk 2]
```diff
@@ -1,79 +1,127 @@
...
-        # manylinux1 offers broader compatibility than manylinux2010 or manylinux2014
-        run: |
-          pip install dist/python_snappy*-manylinux1*.whl
+      - name: Generate source distribution
+        run: python setup.py sdist
 
-      - name: Clean up sources
-        # this cleans up the sources to make sure `import snappy` uses the wheel
-        # an alternative would be to run pytest in a different working directory
-        run: rm -rf setup.py ./snappy
```
Comment (Czaki) [inline .github/workflows/wheel.yml:61]: here is the reason of move to src.

**Comment by Czaki** [comment]
Maybe nice to read: https://iscinumpy.gitlab.io/post/overview-of-cibuildwheel/

In a few days, I will try to use the newest `cibuildwheel` release to build wheels for macOS arm processors.

**Comment by martindurant** [comment]
Thanks @Czaki , and glad to see you still pushing this

**Comment by Czaki** [comment]
It looks ready. Current PR builds 71 wheels:

* Windows 3.5+ CPython (`win32` and `amd64`)  and PyPy
* macOS 2.7, 3.[5-9] `x86_64` and 3.9 `universal2` 
* Linux `x86` and `i686`: 2.7, 3.[5-9] and `aarch64`, `ppc64le` and `s390x`   

list of current wheels: 
```
python_snappy-0.6.0-cp27-cp27m-macosx_10_9_x86_64.whl
python_snappy-0.6.0-cp27-cp27m-manylinux1_i686.whl
python_snappy-0.6.0-cp27-cp27m-manylinux1_x86_64.whl
python_snappy-0.6.0-cp27-cp27m-manylinux2010_i686.whl
python_snappy-0.6.0-cp27-cp27m-manylinux2010_x86_64.whl
python_snappy-0.6.0-cp27-cp27mu-manylinux1_i686.whl
python_snappy-0.6.0-cp27-cp27mu-manylinux1_x86_64.whl
python_snappy-0.6.0-cp27-cp27mu-manylinux2010_i686.whl
python_snappy-0.6.0-cp27-cp27mu-manylinux2010_x86_64.whl
python_snappy-0.6.0-cp35-cp35m-macosx_10_9_x86_64.whl
python_snappy-0.6.0-cp35-cp35m-manylinux1_i686.whl
python_snappy-0.6.0-cp35-cp35m-manylinux1_x86_64.whl
python_snappy-0.6.0-cp35-cp35m-manylinux2010_i686.whl
python_snappy-0.6.0-cp35-cp35m-manylinux2010_x86_64.whl
python_snappy-0.6.0-cp35-cp35m-manylinux2014_aarch64.whl
python_snappy-0.6.0-cp35-cp35m-manylinux2014_ppc64le.whl
python_snappy-0.6.0-cp35-cp35m-manylinux2014_s390x.whl
python_snappy-0.6.0-cp35-cp35m-win32.whl
python_snappy-0.6.0-cp35-cp35m-win_amd64.whl
python_snappy-0.6.0-cp36-cp36m-macosx_10_9_x86_64.whl
python_snappy-0.6.0-cp36-cp36m-manylinux1_i686.whl
python_snappy-0.6.0-cp36-cp36m-manylinux1_x86_64.whl
python_snappy-0.6.0-cp36-cp36m-manylinux2010_i686.whl
python_snappy-0.6.0-cp36-cp36m-manylinux2010_x86_64.whl
python_snappy-0.6.0-cp36-cp36m-manylinux2014_aarch64.whl
python_snappy-0.6.0-cp36-cp36m-manylinux2014_ppc64le.whl
python_snappy-0.6.0-cp36-cp36m-manylinux2014_s390x.whl
python_snappy-0.6.0-cp36-cp36m-win32.whl
python_snappy-0.6.0-cp36-cp36m-win_amd64.whl
python_snappy-0.6.0-cp37-cp37m-macosx_10_9_x86_64.whl
python_snappy-0.6.0-cp37-cp37m-manylinux1_i686.whl
python_snappy-0.6.0-cp37-cp37m-manylinux1_x86_64.whl
python_snappy-0.6.0-cp37-cp37m-manylinux2010_i686.whl
python_snappy-0.6.0-cp37-cp37m-manylinux2010_x86_64.whl
python_snappy-0.6.0-cp37-cp37m-manylinux2014_aarch64.whl
python_snappy-0.6.0-cp37-cp37m-manylinux2014_ppc64le.whl
python_snappy-0.6.0-cp37-cp37m-manylinux2014_s390x.whl
python_snappy-0.6.0-cp37-cp37m-win32.whl
python_snappy-0.6.0-cp37-cp37m-win_amd64.whl
python_snappy-0.6.0-cp38-cp38-macosx_10_9_x86_64.whl
python_snappy-0.6.0-cp38-cp38-manylinux1_i686.whl
python_snappy-0.6.0-cp38-cp38-manylinux1_x86_64.whl
python_snappy-0.6.0-cp38-cp38-manylinux2010_i686.whl
python_snappy-0.6.0-cp38-cp38-manylinux2010_x86_64.whl
python_snappy-0.6.0-cp38-cp38-manylinux2014_aarch64.whl
python_snappy-0.6.0-cp38-cp38-manylinux2014_ppc64le.whl
python_snappy-0.6.0-cp38-cp38-manylinux2014_s390x.whl
python_snappy-0.6.0-cp38-cp38-win32.whl
python_snappy-0.6.0-cp38-cp38-win_amd64.whl
python_snappy-0.6.0-cp39-cp39-macosx_10_9_universal2.whl
python_snappy-0.6.0-cp39-cp39-macosx_10_9_x86_64.whl
python_snappy-0.6.0-cp39-cp39-manylinux1_i686.whl
python_snappy-0.6.0-cp39-cp39-manylinux1_x86_64.whl
python_snappy-0.6.0-cp39-cp39-manylinux2010_i686.whl
python_snappy-0.6.0-cp39-cp39-manylinux2010_x86_64.whl
python_snappy-0.6.0-cp39-cp39-manylinux2014_aarch64.whl
python_snappy-0.6.0-cp39-cp39-manylinux2014_ppc64le.whl
python_snappy-0.6.0-cp39-cp39-manylinux2014_s390x.whl
python_snappy-0.6.0-cp39-cp39-win32.whl
python_snappy-0.6.0-cp39-cp39-win_amd64.whl
python_snappy-0.6.0-pp27-pypy_73-macosx_10_9_x86_64.whl
python_snappy-0.6.0-pp27-pypy_73-manylinux1_x86_64.whl
python_snappy-0.6.0-pp27-pypy_73-manylinux2010_x86_64.whl
python_snappy-0.6.0-pp36-pypy36_pp73-macosx_10_9_x86_64.whl
python_snappy-0.6.0-pp36-pypy36_pp73-manylinux1_x86_64.whl
python_snappy-0.6.0-pp36-pypy36_pp73-manylinux2010_x86_64.whl
python_snappy-0.6.0-pp36-pypy36_pp73-win32.whl
python_snappy-0.6.0-pp37-pypy37_pp73-macosx_10_9_x86_64.whl
python_snappy-0.6.0-pp37-pypy37_pp73-manylinux1_x86_64.whl
python_snappy-0.6.0-pp37-pypy37_pp73-manylinux2010_x86_64.whl
python_snappy-0.6.0-pp37-pypy37_pp73-win32.whl
```

**Comment by martindurant** [comment]
Thanks @Czaki

**Comment by Czaki** [comment]
There is a plan to upload the whole set of wheels in a short time?

**Comment by martindurant** [comment]
If you tell me how! :)

**Comment by Czaki** [comment]
Without a new release, you could download artifacts from the bottom of the build page https://github.com/andrix/python-snappy/actions/runs/572119475 and then use twine to upload (you need to filter out already uploaded wheels)

Zrzut ekranu z 2021-02-16 22-30-53

**Comment by martindurant** [comment]
Lots of wheels uploaded

**Pull request merged**